### ISSUE TITLE
how to compile myself?

### ISSUE THREAD

**Issue by porres** [porres/pd-sfont#27]
I'm now dependant on the help of @Lucarda and @reduzent , I need help learning how to do it myself

**Comment by Lucarda** [comment]
@porres this is already covered in https://github.com/porres/pd-else/tree/master/sfont~

For Windows you will need a real or a VM. I think cross-compiling is not a possibility because you cant get `fluidsynth` as you can with `MSYS2`.

Anyway here it goes:

# Building On Windows With Msys2

Download the 64-bit installer from:

https://www.msys2.org/

Follow installation instructions on: https://github.com/msys2/msys2/wiki/MSYS2-installation

## Open the `MSYS2` shell

Add 32-bit compiler:

`pacman -S mingw32/mingw-w64-i686-gcc`

Add 64-bit compiler:

`pacman -S mingw64/mingw-w64-x86_64-gcc`

Add fluidsynth packages:

`pacman -S mingw32/mingw-w64-i686-fluidsynth`

`pacman -S mingw64/mingw-w64-x86_64-fluidsynth`

Add ntldd packages:

`pacman -S mingw32/mingw-w64-i686-ntldd-git`

`pacman -S mingw64/mingw-w64-x86_64-ntldd-git`

## Open a `MinGW<arch>` shell 

`cd` the ` MinGW<arch>` shell to the downloaded repo and do:

`make`

or you can also specify more options with:

`make PDDIR=<path/to/pd directory>`

then do this command that installs and fills dependencies on the output dir:

`make localdep_windows`

or with more options:

`make localdep_windows PDLIBDIR=<path/you/want-the/output>`

**Comment by porres** [comment]
damn, I don't have a windows machine and was hoping to use a VM... are you sure it's impossible? 

On the other hand, @reduzent uses a Virtual Machine for Linux, right?

**Comment by porres** [comment]
I also need to know the secret of compiling for mac myself without a VM...

**Comment by reduzent** [comment]
> damn, I don't have a windows machine and was hoping to use a VM... are you sure it's impossible?

I think @Lucarda said that you _can_ use a VM (or a real Windows), but _not_ cross-compile for Windows on macOS or Linux. BTW, I also use Windows VM on my Linux to compile for Windows. Simple stuff without dependencies can be cross-compiled for Windows on Linux.

**Comment by reduzent** [comment]
For Darwin-amd64-32 I use this command:

```
make extension=d_amd64 PDDIR=/Users/roman/pd-src/pure-data PDLIBDIR=/Users/roman/pd-src/workspace/Darwin-amd64-32 localdep_macos
```

Tested on macOS Catalina

**Comment by reduzent** [comment]
For Darwin-arm64-32 I use this command:

```
make CPPFLAGS+=-I/opt/homebrew/include LDFLAGS+=-L/opt/homebrew/lib extension=d_arm64 PDDIR=/Users/user/pd-src/pure-data PDLIBDIR=/Users/user/pd-src/workspace/Darwin-arm64-32 localdep_macos
```
NOTE: the CCPFLAGS and LDFLAGS are needed here, because homebrew installs arm64 packages to `/opt/homebrew`. If you have amd64 and thus install amd64 packages, those are probably don't needed. 


Tested on macOS Monterey on a MacBook M1 (You seem to have an Monterey VM with amd64. I don't know how to target arm64 on a amd64 machine. It's probably possible, I just haven't tried myself and I only have an arm64 machine).

**Comment by reduzent** [comment]
Both above commands assume that you have fluidsynth installed with homebrew:  `brew install fluid-synth`

**Comment by Lucarda** [comment]
> I think @Lucarda said that you can use a VM (or a real Windows), but not cross-compile for Windows on macOS or Linux

Yep. That's correct.

**Comment by emviveros** [comment]
I'm starting thinking the possibility to use Docker containers to compile for multiple architectures. The community makes available a lot of images of diverse OSes to compile and crosscompile software. And Docker images are easily usable with github actions if necessary.

Someone have any past experiences with this kind of worflow?

**Comment by porres** [comment]
> Simple stuff without dependencies can be cross-compiled for Windows on Linux.

Yeah, @emviveros helps me with compiling for ELSE and provided me a script that compiles from a VM in Linux to all platforms but macOS. We use a very old Ubuntu and I told him we need to upgrade so we can compile sfont~, which requires cmake >= 3.13 and @reduzent uses Debian Buster with libc=2.28
